src/librustc_mir/borrow_check/mod.rs

```diff
@@ -1242,7 +1242,7 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
         let mut error_reported = false;
         match kind {
             Write(WriteKind::MutableBorrow(BorrowKind::Unique)) => {
-                if let Err(_place_err) = self.is_unique(place) {
+                if let Err(_place_err) = self.is_mutable(place, LocalMutationIsAllowed::Yes) {
```

Comment:
Hmm. I'm not sure if I like this particular commit. In part this is because I don't like the "is local mutation" allowed flag -- I'm trying to put clearly into words why. I guess because it feels like there is context to the `is_mutable` check that is dependent on the path being assigned (and in particular whether it's been assigned before), so `is_mutable` in some sense is not able to answer the question it is being asked.

Reply:
Maybe we could instead refactor `is_mutable` so that it doesn't return a bool but some kind of conditional result.

Reply:
That said, particularly since I can't quite clearly enunciate my concerns, I'm happy enough to land these commits and then take a stab at refactoring things myself -- I might well find I don't like the result.